// @tarojs/helper@4.0.6/dist/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEVICE_RATIO_NAME = exports.taroJsMobxCommon = exports.taroJsMobx = exports.taroJsRedux = exports.taroJsFramework = exports.taroJsQuickAppComponents = exports.taroJsComponents = exports.taroJsMiniComponentsPath = exports.META_TYPE = exports.UPDATE_PACKAGE_LIST = exports.FILE_PROCESSOR_MAP = exports.DEVICE_RATIO = exports.PROJECT_CONFIG = exports.REG_NODE_MODULES_DIR = exports.REG_NODE_MODULES = exports.NODE_MODULES = exports.REG_CSS_IMPORT = exports.REG_TARO_H5_RUNTIME_API = exports.REG_TARO_H5 = exports.REG_TARO_SCOPED_PACKAGE = exports.REG_URL = exports.REG_WXML_IMPORT = exports.REG_TEMPLATE = exports.REG_UX = exports.REG_JSON = exports.REG_FONT = exports.REG_IMAGE = exports.REG_MEDIA = exports.REG_CSS = exports.REG_STYLE = exports.REG_STYLUS = exports.REG_LESS = exports.REG_SASS_SCSS = exports.REG_SASS_SASS = exports.REG_SASS = exports.REG_VUE = exports.REG_SCRIPTS = exports.REG_TYPESCRIPT = exports.REG_SCRIPT = exports.REG_JS = exports.VUE_EXT = exports.SCRIPT_EXT = exports.UX_EXT = exports.TS_EXT = exports.JS_EXT = exports.SCSS_EXT = exports.CSS_EXT = exports.processTypeMap = exports.processTypeEnum = exports.PLATFORMS = void 0;
exports.defaultMainFields = exports.FRAMEWORK_MAP = exports.ENTRY = exports.NPM_DIR = exports.TEMP_DIR = exports.SOURCE_DIR = exports.OUTPUT_DIR = exports.TARO_GLOBAL_CONFIG_FILE = exports.TARO_GLOBAL_CONFIG_DIR = exports.TARO_BASE_CONFIG = exports.TARO_CONFIG_FOLDER = exports.DEFAULT_TEMPLATE_SRC_GITEE = exports.DEFAULT_TEMPLATE_SRC = exports.isWindows = void 0;
const os = __importStar(require("node:os"));
const terminal_1 = require("./terminal");
// eslint-disable-next-line dot-notation
exports.PLATFORMS = (global['PLATFORMS'] = global['PLATFORMS'] || {});
var processTypeEnum;
(function (processTypeEnum) {
    processTypeEnum["START"] = "start";
    processTypeEnum["CREATE"] = "create";
    processTypeEnum["COMPILE"] = "compile";
    processTypeEnum["CONVERT"] = "convert";
    processTypeEnum["COPY"] = "copy";
    processTypeEnum["GENERATE"] = "generate";
    processTypeEnum["MODIFY"] = "modify";
    processTypeEnum["ERROR"] = "error";
    processTypeEnum["WARNING"] = "warning";
    processTypeEnum["UNLINK"] = "unlink";
    processTypeEnum["REFERENCE"] = "reference";
    processTypeEnum["REMIND"] = "remind";
})(processTypeEnum || (exports.processTypeEnum = processTypeEnum = {}));
exports.processTypeMap = {
    ["create" /* processTypeEnum.CREATE */]: {
        name: '创建',
        color: 'cyan',
    },
    ["compile" /* processTypeEnum.COMPILE */]: {
        name: '编译',
        color: 'green',
    },
    ["convert" /* processTypeEnum.CONVERT */]: {
        name: '转换',
        color: terminal_1.chalk.rgb(255, 136, 0),
    },
    ["copy" /* processTypeEnum.COPY */]: {
        name: '拷贝',
        color: 'magenta',
    },
    ["generate" /* processTypeEnum.GENERATE */]: {
        name: '生成',
        color: 'blue',
    },
    ["modify" /* processTypeEnum.MODIFY */]: {
        name: '修改',
        color: 'yellow',
    },
    ["error" /* processTypeEnum.ERROR */]: {
        name: '错误',
        color: 'red',
    },
    ["warning" /* processTypeEnum.WARNING */]: {
        name: '警告',
        color: 'yellowBright',
    },
    ["unlink" /* processTypeEnum.UNLINK */]: {
        name: '删除',
        color: 'magenta',
    },
    ["start" /* processTypeEnum.START */]: {
        name: '启动',
        color: 'green',
    },
    ["reference" /* processTypeEnum.REFERENCE */]: {
        name: '引用',
        color: 'blue',
    },
    ["remind" /* processTypeEnum.REMIND */]: {
        name: '提示',
        color: 'green',
    },
};
exports.CSS_EXT = ['.css', '.scss', '.sass', '.less', '.styl', '.stylus', '.wxss', '.acss'];
exports.SCSS_EXT = ['.scss'];
exports.JS_EXT = ['.js', '.jsx'];
exports.TS_EXT = ['.ts', '.tsx'];
exports.UX_EXT = ['.ux'];
exports.SCRIPT_EXT = exports.JS_EXT.concat(exports.TS_EXT);
exports.VUE_EXT = ['.vue'];
exports.REG_JS = /\.m?js(\?.*)?$/;
exports.REG_SCRIPT = /\.m?(js|jsx)(\?.*)?$/;
exports.REG_TYPESCRIPT = /\.(tsx|ts)(\?.*)?$/;
exports.REG_SCRIPTS = /\.m?[tj]sx?$/i;
exports.REG_VUE = /\.vue$/i;
exports.REG_SASS = /\.(s[ac]ss)$/;
exports.REG_SASS_SASS = /\.sass$/;
exports.REG_SASS_SCSS = /\.scss$/;
exports.REG_LESS = /\.less$/;
exports.REG_STYLUS = /\.styl(us)?$/;
exports.REG_STYLE = /\.(css|scss|sass|less|styl|stylus|wxss|acss|ttss|jxss|qss)(\?.*)?$/;
exports.REG_CSS = /\.(css|qss|jxss|wxss|acss|ttss)(\?.*)?$/;
exports.REG_MEDIA = /\.(mp4|webm|ogg|mp3|m4a|wav|flac|aac)(\?.*)?$/;
exports.REG_IMAGE = /\.(png|jpe?g|gif|bpm|svg|webp)(\?.*)?$/;
exports.REG_FONT = /\.(woff2?|eot|ttf|otf)(\?.*)?$/;
exports.REG_JSON = /\.json(\?.*)?$/;
exports.REG_UX = /\.ux(\?.*)?$/;
exports.REG_TEMPLATE = /\.(wxml|axml|ttml|qml|swan|jxml)(\?.*)?$/;
exports.REG_WXML_IMPORT = /<import(.*)?src=(?:(?:'([^']*)')|(?:"([^"]*)"))/gi;
exports.REG_URL = /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z0-9\u00a1-\uffff][a-z0-9\u00a1-\uffff_-]{0,62})?[a-z0-9\u00a1-\uffff]\.)+(?:[a-z\u00a1-\uffff]{2,}\.?))(?::\d{2,5})?(?:[/?#]\S*)?$/i;
exports.REG_TARO_SCOPED_PACKAGE = /@tarojs[\\/][a-z]+/;
exports.REG_TARO_H5 = /taro-h5[\\/]dist[\\/](api[\\/]taro|index\.esm)/;
exports.REG_TARO_H5_RUNTIME_API = /@tarojs[\\/]plugin-platform-h5[\\/]dist[\\/]runtime[\\/]apis[\\/]index/;
exports.REG_CSS_IMPORT = /@import (["'])(.+?)\1;/g;
exports.NODE_MODULES = 'node_modules';
exports.REG_NODE_MODULES = /node_modules/;
exports.REG_NODE_MODULES_DIR = /[\\/]node_modules[\\/]/gi;
exports.PROJECT_CONFIG = 'config/index';
exports.DEVICE_RATIO = {
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2,
};
exports.FILE_PROCESSOR_MAP = {
    '.js': 'babel',
    '.scss': 'sass',
    '.sass': 'sass',
    '.less': 'less',
    '.styl': 'stylus',
};
exports.UPDATE_PACKAGE_LIST = [
    'babel-plugin-transform-react-jsx-to-rn-stylesheet',
    'taro-css-to-react-native',
    'stylelint-config-taro-rn',
    'stylelint-taro-rn',
    'babel-plugin-transform-taroapi',
    'babel-preset-taro',
    'eslint-config-taro',
    'postcss-html-transform',
    'postcss-plugin-constparse',
    'postcss-pxtransform',
    '@tarojs/shared',
    '@tarojs/taro',
    '@tarojs/cli',
    '@tarojs/api',
    '@tarojs/components',
    '@tarojs/components-advanced',
    '@tarojs/components-react',
    '@tarojs/components-rn',
    '@tarojs/extend',
    '@tarojs/taro-h5',
    '@tarojs/taro-rn',
    '@tarojs/rn-runner',
    '@tarojs/rn-style-transformer',
    '@tarojs/rn-supporter',
    '@tarojs/rn-transformer',
    '@tarojs/helper',
    '@tarojs/taro-loader',
    '@tarojs/react',
    '@tarojs/plugin-framework-react',
    '@tarojs/plugin-framework-vue3',
    '@tarojs/plugin-framework-solid',
    '@tarojs/plugin-react-devtools',
    '@tarojs/plugin-vue-devtools',
    '@tarojs/router',
    '@tarojs/router-rn',
    '@tarojs/runner-utils',
    '@tarojs/runtime',
    '@tarojs/runtime-rn',
    '@tarojs/service',
    '@tarojs/with-weapp',
    '@tarojs/taroize',
    '@tarojs/plugin-inject',
    '@tarojs/plugin-platform-weapp',
    '@tarojs/plugin-platform-alipay',
    '@tarojs/plugin-platform-swan',
    '@tarojs/plugin-platform-tt',
    '@tarojs/plugin-platform-qq',
    '@tarojs/plugin-platform-jd',
    '@tarojs/plugin-platform-h5',
    '@tarojs/plugin-platform-harmony-ets',
    '@tarojs/plugin-platform-harmony-hybrid',
    '@tarojs/plugin-html',
    '@tarojs/plugin-mini-ci',
    '@tarojs/plugin-http',
    '@tarojs/webpack5-runner',
    '@tarojs/webpack5-prebundle',
    '@tarojs/vite-runner',
    '@tarojs/create-app',
    '@tarojs/cli-convertor',
    '@tarojs/transformer-wx',
];
var META_TYPE;
(function (META_TYPE) {
    META_TYPE["ENTRY"] = "ENTRY";
    META_TYPE["PAGE"] = "PAGE";
    META_TYPE["COMPONENT"] = "COMPONENT";
    META_TYPE["NORMAL"] = "NORMAL";
    META_TYPE["STATIC"] = "STATIC";
    META_TYPE["CONFIG"] = "CONFIG";
    META_TYPE["EXPORTS"] = "EXPORTS";
})(META_TYPE || (exports.META_TYPE = META_TYPE = {}));
exports.taroJsMiniComponentsPath = '@tarojs/components/mini';
exports.taroJsComponents = '@tarojs/components';
exports.taroJsQuickAppComponents = '@tarojs/components-qa';
exports.taroJsFramework = '@tarojs/taro';
exports.taroJsRedux = '@tarojs/redux';
exports.taroJsMobx = '@tarojs/mobx';
exports.taroJsMobxCommon = '@tarojs/mobx-common';
exports.DEVICE_RATIO_NAME = 'deviceRatio';
exports.isWindows = os.platform() === 'win32';
exports.DEFAULT_TEMPLATE_SRC = 'github:NervJS/taro-project-templates#v4.0';
exports.DEFAULT_TEMPLATE_SRC_GITEE = 'direct:https://gitee.com/o2team/taro-project-templates.git#v4.0';
exports.TARO_CONFIG_FOLDER = '.taro4.0';
exports.TARO_BASE_CONFIG = 'index.json';
exports.TARO_GLOBAL_CONFIG_DIR = '.taro-global-config';
exports.TARO_GLOBAL_CONFIG_FILE = 'index.json';
exports.OUTPUT_DIR = 'dist';
exports.SOURCE_DIR = 'src';
exports.TEMP_DIR = '.temp';
exports.NPM_DIR = 'npm';
exports.ENTRY = 'app';
var FRAMEWORK_MAP;
(function (FRAMEWORK_MAP) {
    FRAMEWORK_MAP["VUE3"] = "vue3";
    FRAMEWORK_MAP["REACT"] = "react";
    FRAMEWORK_MAP["Solid"] = "solid";
})(FRAMEWORK_MAP || (exports.FRAMEWORK_MAP = FRAMEWORK_MAP = {}));
exports.defaultMainFields = ['browser', 'module', 'jsnext:main', 'main'];
